openedx/core/djangoapps/course_groups/cohorts.py

```diff
@@ -423,28 +438,63 @@ def add_user_to_cohort(cohort, username_or_email):
         username_or_email: string.  Treated as email if has '@'
 
     Returns:
-        Tuple of User object and string (or None) indicating previous cohort
+        User object, string (or None) indicating previous cohort, and whether the user is a preassigned user or not
 
     Raises:
-        User.DoesNotExist if can't find user.
+        User.DoesNotExist if can't find user. However, if a valid email is provided for the user, it is stored
+        in a database so that the user can be added to the cohort if they eventually enroll in the course.
         ValueError if user already present in this cohort.
+        ValidationError if an invalid email address is entered.
+        User.DoesNotExist if a user could not be found.
     """
-    user = get_user_by_username_or_email(username_or_email)
+    try:
+        user = get_user_by_username_or_email(username_or_email)
 
-    membership = CohortMembership(course_user_group=cohort, user=user)
-    membership.save()  # This will handle both cases, creation and updating, of a CohortMembership for this user.
+        membership = CohortMembership(course_user_group=cohort, user=user)
+        membership.save()  # This will handle both cases, creation and updating, of a CohortMembership for this user.
 
-    tracker.emit(
-        "edx.cohort.user_add_requested",
-        {
-            "user_id": user.id,
-            "cohort_id": cohort.id,
-            "cohort_name": cohort.name,
-            "previous_cohort_id": membership.previous_cohort_id,
-            "previous_cohort_name": membership.previous_cohort_name,
-        }
-    )
-    return (user, membership.previous_cohort_name)
+        tracker.emit(
+            "edx.cohort.user_add_requested",
+            {
+                "user_id": user.id,
+                "cohort_id": cohort.id,
+                "cohort_name": cohort.name,
+                "previous_cohort_id": membership.previous_cohort_id,
+                "previous_cohort_name": membership.previous_cohort_name,
+            }
+        )
+        return (user, membership.previous_cohort_name, False)
+    except User.DoesNotExist as ex:
+        # If username_or_email is an email address, store in database.
+        try:
+            validate_email(username_or_email)
+
+            try:
+                user = UnregisteredLearnerCohortAssignments.objects.get(
```

Comment:
As mentioned below in the test, it would be good to have coverage of this case (updating the cohort assignment).